src/client/common/utils/async.ts

```diff
@@ -237,3 +237,25 @@ export async function flattenIterator<T>(iterator: AsyncIterator<T, void>): Prom
     }
     return results;
 }
+
+export class ChainedExecutions<T> {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private pendingCellExecution: Promise<any> = Promise.resolve();
+    /**
+     * Clears the chained promises.
+     */
+    public clear() {
+        this.pendingCellExecution = Promise.resolve();
+    }
+    public async chainExecution(next: () => Promise<T>): Promise<T> {
+        const deferred = createDeferred<T>();
+        const chainedPromise = this.pendingCellExecution.then(() => {
+            const nextPromise = next();
+            nextPromise.then((result) => deferred.resolve(result)).catch((ex) => deferred.reject(ex));
+            return nextPromise;
+        });
+        this.pendingCellExecution = chainedPromise;
+
+        return Promise.race([deferred.promise, chainedPromise]);
```

Comment:
Either return when the new succeeds or the chain get rejected.

Reply:
This seems rather complicated. Could you do this instead?

```
let pendingError = false;
const pending = this.pending ?? Promise.resolve():
this.pending = pending.then(next).catch((e) => { if pendingError { throw e; } return next(); });
return this.pending;
```

That should:
1) Chain the promises
1) Only throw a pending exception once

I believe that's the behavior we want? 

I think your new version does not work if pending fails.

Reply:
Ah I see you require the 'clear' to be called. That is kinda sucky. It woudl be better if we could eliminate that interdependency.

Reply:
Sure, will try that.

Reply:
I see a problem;
> this.pending = pending.then(next).catch((e) => { if pendingError { throw e; } pendingError = false; return next(); });

Assume previous pending failed.
Now it will go into the catch block and the exception is swallowed, because of the `catch` clause.
So when we chain a new proimise on top of this, the previous promise failed, & then due to the chaining we ignore that & continue to the next promise.

Ideally the whole thing should just fail. Else its not a chained. Its more like throw once & continue.

Reply:
Basically we want everything that gets chained to that to stop running

Reply:
Feel like we need a separate class that manages the batch of executions. If one thing fails, then everything in that batch shoulld stop (single thing should shortcircuit the whole operation).

This way if we have a batch of executions already running, just add the new cell to the existing batch. & if one cell fails (errors for kernel or what ever reason), then stop all.

& if a batch doesn't exist, then create a new batch.

& once that batch completes, then tear down that class & dispose everything in it.

Personally I prefer this option, not sure if its over kill.
Today we kinda do that by using the `stack` in `this.stackOfCellsToExecuteByDocument`. If we have an array with items, this means we have an execution going on & we just append the new execution to that job.

@IanMatthewHuff  @rchiodo thoughts?
Or should we just meet up and discuss this.

Reply:
At the same time i do'nt want to over complicate it, but would also like to simplify it..
question is, is the current code clear enough...

Reply:
I believe this is what we want. 

>  Else its not a chained. Its more like throw once & continue

Well the reason for that was the caller would keep getting the same error over and over. It would never retry. Maybe your clear idea would be better then?

Reply:
So then we could just have this:

```
const pending = this.pending ?? Promise.resolve();
this.pending = pending.then(next);
return this.pending;
```

and then do the clear like you have now?